hydra/core/plugins.py

```diff
@@ -103,6 +106,57 @@ def is_in_toplevel_plugins_module(clazz: str) -> bool:
             "hydra._internal.core_plugins."
         )
 
+    @staticmethod
+    def _setup_plugin(
+        plugin: Any,
+        task_function: TaskFunction,
+        config: DictConfig,
+        config_loader: Optional[ConfigLoader] = None,
+        hydra_context: Optional[HydraContext] = None,
+    ) -> Any:
+        """
+        With HydraContext introduced in #1581, we need to set up the plugins in a way
+        that's compatible with both Hydra 1.0 and Hydra 1.1 syntax.
+        This method should be deleted in the next major release.
+        """
+        assert isinstance(plugin, Sweeper) or isinstance(plugin, Launcher)
+        assert (
+            config_loader is not None or hydra_context is not None
+        ), "config_loader and hydra_context cannot both be None"
+
+        param_keys = signature(plugin.setup).parameters.keys()
+
+        if "hydra_context" not in param_keys:
+            warnings.warn(
+                message=(
+                    "\n"
+                    "\tPlugin's setup() signature has changed in Hydra 1.1.\n"
+                    "\tSupport for the old style will be removed in Hydra 1.2.\n"
+                    "\tFor more info, check https://github.com/facebookresearch/hydra/pull/1581.\n"
+                ),
+                category=UserWarning,
+            )
+            config_loader = (
+                config_loader
+                if config_loader is not None
+                else hydra_context.config_loader  # type: ignore
+            )
+            plugin.setup(  # type: ignore
+                config=config,
+                config_loader=config_loader,
+                task_function=task_function,
+            )
+        else:
+            if hydra_context is None:
+                assert config_loader is not None
+                hydra_context = HydraContext(
+                    config_loader=config_loader, callbacks=Callbacks()
+                )
```

Comment:
when is hydra_context none here?
If this is just to support tests, have them instantiate the HydraContext.

Reply:
`hydra_context` could be `None` in the case of an outdated Sweeper tried to instantiate a BasicLauncher. 

I've added a comment here. 

```
u$ python /Users/jieru/workspace/hydra-fork/hydra/plugins/hydra_ax_sweeper/example/banana.py -m
/Users/jieru/workspace/hydra-fork/hydra/hydra/core/plugins.py:131: UserWarning:


Plugin's setup() signature has changed in Hydra 1.1.
Support for the old style will be removed in Hydra 1.2.
For more info, check https://github.com/facebookresearch/hydra/pull/1581.


[2021-05-06 10:38:30,027][HYDRA] AxSweeper is optimizing the following parameters: 
banana.x: range=[-5, 5]
banana.y: range=[-5.0, 10.1]
[INFO 05-06 10:38:30] ax.service.utils.instantiation: Inferred value type of ParameterType.INT for parameter banana.x. If that is not the expected value type, you can explicity specify 'value_type' ('int', 'float', 'bool' or 'str') in parameter dict.
[INFO 05-06 10:38:30] ax.service.utils.instantiation: Inferred value type of ParameterType.FLOAT for parameter banana.y. If that is not the expected value type, you can explicity specify 'value_type' ('int', 'float', 'bool' or 'str') in parameter dict.
[INFO 05-06 10:38:30] ax.modelbridge.dispatch_utils: Using Bayesian Optimization generation strategy: GenerationStrategy(name='Sobol+GPEI', steps=[Sobol for 5 trials, GPEI for subsequent trials]). Iterations after 5 will take longer to generate due to  model-fitting.
[2021-05-06 10:38:30,130][HYDRA] AxSweeper is launching 5 jobs
Traceback (most recent call last):
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/utils.py", line 212, in run_and_report
    return func()
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/utils.py", line 379, in <lambda>
    lambda: hydra.multirun(
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/hydra.py", line 139, in multirun
    ret = sweeper.sweep(arguments=task_overrides)
  File "/Users/jieru/opt/anaconda3/envs/pytest38/lib/python3.8/site-packages/hydra_plugins/hydra_ax_sweeper/ax_sweeper.py", line 29, in sweep
    return self.sweeper.sweep(arguments)
  File "/Users/jieru/opt/anaconda3/envs/pytest38/lib/python3.8/site-packages/hydra_plugins/hydra_ax_sweeper/_core.py", line 180, in sweep
    self.sweep_over_batches(
  File "/Users/jieru/opt/anaconda3/envs/pytest38/lib/python3.8/site-packages/hydra_plugins/hydra_ax_sweeper/_core.py", line 217, in sweep_over_batches
    rets = self.launcher.launch(
  File "/Users/jieru/workspace/hydra-fork/hydra/hydra/_internal/core_plugins/basic_launcher.py", line 55, in launch
    assert self.hydra_context is not None
AssertionError

```

Reply:
gotcha.